# FSU-Hackathon. 

My name is Eric Dolores, I belong to Chihack night which is a weekly hackathon in Chicago. In Chihack people gather to use their own expertice to solve the problems of the city. In my first project I did webscrapping and in the second one I helped an artist automatize the production of his sculptures. This collaboration lead to us participating and winning an international art contest in Germany.
 
My first contribution to FSU-hackathon will be to introduce the basics of Python 3 to the community. 

# Requirements.

Install Anaconda: https://www.anaconda.com/download/

We will use the book: https://runestone.academy/runestone/static/FSUhackathon/index.html please enroll there as I can suggest you particular exercises based on your experience.

I will assume all of you have some experience programming. 


Note:
If you are already comfortable with the basics of Python, use this session to find what is faster in Python: to multiply a number by 2 or to shift it. Don't google the answer!

 # "The devil is in the details."

A variable is a sticky note that you put on objects. To assign a value to a variable we write to the left of  "=", a variable and to the right of "=" an expression which is evaluated by the Python interpreter and then assigned to the name. The difference between the left and right hand sides of the assignment statement is often confusing to new programmers. In the following assignment:

    n = n + 1

n plays a very different role on each side of the =. On the right it is a value and makes up part of the expression which will be evaluated by the Python interpreter before assigning it to the name on the left.


In [ ]:
minutes = 645
today = "Tuesday"

print(today)

Operators are special tokens that represent computations like addition, multiplication and division. The values the operator works on are called operands.

In [ ]:
print(2 + 3)
print(2 - 3)
print(2 * 3)
print(2 ** 3)
print(3 ** 2)


When a variable name appears in the place of an operand, it is replaced with the value that it refers to before the operation is performed. For example, what if we wanted to convert 645 minutes into hours. In Python 3, division is denoted by the operator token / which always evaluates to a floating point result.

In [ ]:
minutes = 645
hours = minutes / 60
print(hours)


Exercise: What value is printed when the following statements executes?

In [1]:
x = 12
x = x - 1
print(x)
  

11


In [ ]:
y = 2
y+1 = 2
print(y)

What if, on the other hand, we had wanted to know how many whole hours there are and how many minutes remain. To help answer this question, Python gives us a second flavor of the division operator. This version, called integer division, uses the token //. It always truncates its result down to the next smallest integer (to the left on the number line).

In [2]:
print(7 / 4)
print(7 // 4)
minutes = 645
hours = minutes // 60
print(hours)

1.75
1
10


Notice that the result of floating point division is 1.75 but the result of the integer division is simply 1. Take care that you choose the correct flavor of the division operator. If you’re working with expressions where you need floating point values, use the division operator /. If you want an integer result, use //.

The modulus operator, sometimes also called the remainder operator or integer remainder operator works on integers (and integer expressions) and yields the remainder when the first operand is divided by the second. In Python, the modulus operator is a percent sign (%)

In [3]:
quotient = 7 // 3     # This is the integer division operator
print(quotient)
remainder = 7 % 3
print(remainder)


2
1


Exercise: What value is printed when the following statement executes?




In [ ]:
print(18 / 4)


Remember that a variable is a sticky note that you put on objects.


It is important to note that in mathematics, a statement of equality is always true. If a is equal to b now, then a will always equal to b. In Python, an assignment statement can make two variables refer to the same object and therefore have the same value. They appear to be equal. However, because of the possibility of reassignment, they don’t have to stay that way:

In [1]:
a = 5
b = a    # after executing this line, a and b are now equal
print(a, b)

a = 3    # after executing this line, a and b are no longer equal
print(a, b)

5 5
3 5


Note:

In some programming languages, a different symbol is used for assignment, such as <- or :=. The intent is that this will help to avoid confusion. Python chose to use the tokens = for assignment, and == for equality. This is a popular choice also found in languages like C, C++, Java, and C#.


Exercise: After the following statements, what are the values of x and y?

In [ ]:
x = 15
y = x
x = 22
print(x, y)

Reassignment can be used when the new value of the variable depends on the old. For example,

In [ ]:
x = 7
x = x + 1

Once in a while, it is useful to swap the values of two variables. With conventional assignment statements, we have to use a temporary variable. For example, to swap a and b:

In [ ]:
a = 1
b = 12

temp = a
a = b
b = temp

print(a,b)

We will see tuples in the future, at this moment just keep in mind the pythonist way to swap:

In [ ]:
a = 1
b = 12

a, b = b, a

print(a,b)

# Homework:
Do the last two exercises of every section until you get stuck. Then do the first 4 exercises of that section.
    https://runestone.academy/runestone/static/FSUhackathon/index.html

Answer to intial question: In Python, is it faster to multiply a number x by 2 or to shift the number to the left?
https://stackoverflow.com/questions/37053379/times-two-faster-than-bit-shift